### PR TITLE
[Enhance] Tweaks on Memory Usage and Performance

### PR DESCRIPTION
**Pull request by berniechiu**
安安 未來新同事早~
年前看一下 Open Source 來發個 PR 哈哈

## Objective

Little tweaks on Rails memory usage and performance

## Summary

- Upgrade ruby minor version
- Remove unused Rails modules and gems
- Test spec fixed
- Events fixed and optimized

Actually I'm not sure this is a proper fix or not since the original intention in commit message is

The old methods actually still loop through all the events actually, so I change to a more proper use of `uniq` with `block` condition
```ruby
# [Old] Same replay token from events still being selected
@uniq_event_tokens = params['events'].map { |e| e['replyToken'] }.uniq
@events = params['events'].select { |e| @uniq_event_tokens.include?(e['replyToken']) }

# [New] Duplicate reply token only selected once and reduce the extra loop
@events = params[:events].uniq { |event| event['replyToken'] }
```

Results:
```
Warming up --------------------------------------
            original     8.408k i/100ms
                 new    14.777k i/100ms
Calculating -------------------------------------
            original     82.067k (± 9.0%) i/s -    411.992k in   5.066976s
                 new    174.802k (± 1.0%) i/s -    886.620k in   5.072660s

Comparison:
                 new:   174802.1 i/s
            original:    82066.6 i/s - 2.13x  slower
```

- Remove the multiple object initializations with LineBot API Client

### CODE REVIEW

File: app/controllers/listener/lines_controller.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 class Listener::LinesController < ApplicationController
-  before_action :event, only: [:check]
+  before_action :set_event, only: [:check]
```
Comment (CarolHsu) [inline app/controllers/listener/lines_controller.rb:2]: I try not to use prefix `get_` or `set_` recently, but not a big deal 👌

File: app/controllers/listener/lines_controller.rb [hunk 2]
```diff
@@ -16,9 +16,8 @@ def forwardable?(rumor)
     rumor.length > 20
   end
 
-  def event
-    @uniq_event_tokens = params['events'].map { |e| e['replyToken'] }.uniq
-    @events = params['events'].select { |e| @uniq_event_tokens.include?(e['replyToken']) }
+  def set_events
+    @events = params[:events].uniq { |event| event['replyToken'] }
```
Comment (CarolHsu) [inline app/controllers/listener/lines_controller.rb:20]: 👍 👍 👍

File: app/lib/intro.rb [hunk 3]
```diff
@@ -1,21 +1,11 @@
-require 'line/bot'
-
 class Intro
   class << self
     def talk(token)
-      initiate_client
-      @client.reply_message(token, introduction)
+      LineBotClient.reply_message(token, introduction)
```
Comment (CarolHsu) [inline app/lib/intro.rb:4]: nice, I'd like to do this these couple day and you did!!!!!
sweet <3

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -24,12 +24,12 @@
 
 # Checks for pending migrations and applies them before tests are run.
 # If you are not using ActiveRecord, you can remove these lines.
-begin
-  ActiveRecord::Migration.maintain_test_schema!
-rescue ActiveRecord::PendingMigrationError => e
-  puts e.to_s.strip
-  exit 1
-end
+# begin
```
Comment (CarolHsu) [inline spec/rails_helper.rb:27]: also in my wish list <3 <3 <3

**Review by CarolHsu — APPROVED** [review]
thanks for this PR, resolve 2 things I'd like to do (once I have any chance)!

I'm also thinking about how to do statistics / analytics, since this PR removed active record related code (not a bad idea), I think I would do it by google analytics later on.

File: app/controllers/listener/lines_controller.rb (hunk 1 above)
Comment (berniechiu) [reply app/controllers/listener/lines_controller.rb:2]: oh you can change it anyway you like, haha. I just find most the time the function name are verbs

**Pull request merged**